src/main/webapp/app/course/competencies/competency-accordion/competency-accordion.component.ts

```diff
@@ -0,0 +1,167 @@
+            .flatMap((lectureUnitId) => this.metrics.lectureUnitStudentMetricsDTO?.lectureUnitInformation?.[lectureUnitId] ?? [])
+            .filter((lectureUnit) => dayjs(lectureUnit.releaseDate).isBefore(dayjs()))
+            .sort((a, b) => dayjs(a.releaseDate).diff(dayjs(b.releaseDate)));
+    }
+
+    toggle() {
+        this.open = !this.open;
+        this.accordionToggle.emit({ opened: this.open, index: this.index });
+    }
+
+    get jolRating() {
+        return this.metrics.competencyMetrics?.jolValues?.[this.competency.id];
+    }
+
+    onRatingChange(newRating: number) {
+        if (this.metrics.competencyMetrics) {
+            this.metrics.competencyMetrics.jolValues = {
+                ...this.metrics.competencyMetrics.jolValues,
+                [this.competency.id]: newRating,
+            };
+        }
+    }
+
+    getUserProgress(): CompetencyProgress {
+        const progress = this.metrics.competencyMetrics?.progress?.[this.competency.id] ?? 0;
+        const confidence = this.metrics.competencyMetrics?.confidence?.[this.competency.id] ?? 0;
+        return { progress, confidence } as CompetencyProgress;
+    }
+
+    get progress() {
+        if (this.jolRating === undefined) {
+            return 0;
+        }
+        return this.getProgress(this.getUserProgress());
+    }
+
+    get lectureUnitsProgress() {
+        if (!this.metrics.lectureUnitStudentMetricsDTO) {
+            return undefined;
+        }
+
+        const competencyLectureUnits = this.metrics.competencyMetrics?.lectureUnits?.[this.competency.id];
+        const completedLectureUnits = competencyLectureUnits?.filter((lectureUnitId) => this.metrics.lectureUnitStudentMetricsDTO?.completed?.includes(lectureUnitId)).length ?? 0;
+        if (!competencyLectureUnits) {
+            return undefined;
+        }
+        const progress = (completedLectureUnits / competencyLectureUnits.length) * 100;
+        return round(progress, 1);
+    }
+
+    get exercisesProgress() {
+        if (!this.metrics.exerciseMetrics) {
+            return undefined;
+        }
+
+        const competencyExercises = this.metrics.competencyMetrics?.exercises?.[this.competency.id];
+        const completedExercises = competencyExercises?.filter((exerciseId) => this.metrics.exerciseMetrics?.completed?.includes(exerciseId)).length ?? 0;
+        if (!competencyExercises) {
+            return undefined;
+        }
+        const progress = (completedExercises / competencyExercises.length) * 100;
+        return round(progress, 1);
+    }
+
+    get confidence() {
+        if (this.jolRating === undefined) {
+            return 0;
+        }
+        return this.getConfidence(this.getUserProgress(), this.competency.masteryThreshold!);
+    }
+
+    get mastery() {
+        if (this.jolRating === undefined) {
+            return 0;
+        }
+        return this.getMastery(this.getUserProgress(), this.competency.masteryThreshold!);
+    }
```

Comment:
@FelixTJDietrich this component uses very many getters which are then used in the template. These are still function calls, which we want to avoid in the template. Please use normal variables or signals.